Allow the user to choose an export folder during each export

mcilroyc opened the pull request:
This PR contains two major things:

1. When a user initiates an export, they are presented with a "folder chooser" dialog to choose where the assets should be exported.  This starts in their home directory, but then should remember their choice during this session.  (improvements to follow).
1. Consistency with the new version of the generator-spaces plugin that went through a few breaking changes in order to clean things up.

Requires generator-spaces PR
merged ~~https://github.com/adobe-photoshop/generator-spaces/pull/3~~
min version:  jenkins playground-dev builds (~535win / 538mac) 

One particular caveat:
When in the OS folder chooser dialog, do NOT try to create a new folder.  Choose an existing one.

volfied commented on src/js/actions/export.js:
```diff
@@ -55,6 +55,8 @@ define(function (require, exports) {
      */
     var _exportService;
 
+    var _lastFolderPath;
```
Need JSdoc for this

shaoshing replied:
We should definitely make JSCS checking jsdoc existence for us.

volfied replied:
JSCS `jsDoc#enforceExistence` only applies to functions :/

volfied commented on src/js/actions/export.js:
```diff
@@ -438,6 +440,13 @@ define(function (require, exports) {
     setAllNonABLayersExportEnabled.writes = [];
     setAllNonABLayersExportEnabled.transfers = [setLayerExportEnabled];
 
+    var promptForFolder = function () {
```
Ditto.

volfied commented on src/js/actions/export.js:
```diff
@@ -461,16 +470,30 @@ define(function (require, exports) {
+                function (err) {
+                    // promptForFolder rejected, probably just user-canceled
+                    if (err.message && err.message.startsWith("cancelError: cancel")) {
+                        log.warn("Prompt for folder failed, and it wasn't a simple 'cancel'");
```
you say it wasn't a simple cancel, but this boolean check is testing for `startsWith("cancelError")`

mcilroyc replied:
Yeah... this is a little janky, but I think it is an acceptable hack for now. This has to do with the way generator "Domain" infrastructure is bundling the error.  I didn't want to dig too deep since this implementation is still somewhat tentative.

volfied commented on src/js/actions/export.js:
```diff
@@ -461,16 +470,30 @@ define(function (require, exports) {
+                    if (err.message && err.message.startsWith("cancelError: cancel")) {
+                        log.warn("Prompt for folder failed, and it wasn't a simple 'cancel'");
+                    }
+                    return Promise.resolve();
```
Maybe resolve to an empty array here would make the function more consistent.

mcilroyc replied:
I feel that since the JSDoc makes no mention of what is being promised, it is OK as is.  I updated the JSDoc to be slightly more clear.  This promise gets dumped on the floor anyway.

volfied commented:
Did one pass, initially looks good except couple nits and missing JSDoc.

mcilroyc commented:
@volfied I fixed up some JSDoc, and answered some other concerns (hopefully).  I don't want to merge this yet ... but I DO want to get it to the point that it can be merged in close concert with my pushing the generator plugin to perforce (so that our playground-dev jenkins builds are not DOA for very long).  So feel free to keep poking holes in this as you see fit.

iwehrman commented on src/js/actions/export.js:
```diff
@@ -461,16 +481,30 @@ define(function (require, exports) {
+                },
+                function (err) {
+                    // promptForFolder rejected, probably just user-canceled
+                    if (err.message && err.message.startsWith("cancelError: cancel")) {
```
Shouldn't you re-throw the error the error in this case?

mcilroyc replied:
well, I intentionally was not re-throwing, but now that I'm staring at it again I can see how it would be safer to be strict about this.  Will change.

iwehrman commented on src/js/actions/export.js:
```diff
@@ -439,10 +448,21 @@ define(function (require, exports) {
+     * Open a dialog to allow the user to choose a directory
+     * @return {Promise.<string>} Resolves with pathName, or rejects if user-canceled
+     */
+    var promptForFolder = function () {
```
I can't find any place where this action is used. Am I just missing it?

mcilroyc replied:
derp.  I think I originally was orchestrating this from outside the action (JSX), but now I'm just calling _exportService directly in `exportAllAssets`.  I'll remove this as I don't see any reason we would want to transfer to it right now.

iwehrman commented:
Just a couple Qs. Looks good to me otherwise!

mcilroyc commented:
Ian's nits addressed, and squashed into the previous "review comments" commit.  This is, afaict, ready to merge. I updated the PR description with the latest information.

iwehrman commented:
One more wish: file a bug for:
> When in the OS folder chooser dialog, do NOT try to create a new folder. Choose an existing one.
Otherwise, looks good!
